torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -298,22 +299,25 @@ Node* CloneNodeToGraph(
             return input;
           }
           default: {
+            // Try to lookup input value and insert it into the graph.
+            // If the input value is unknown, set it to graph input in the new
+            // graph, and copy over metadata, such as datatype and shape.
+            ::c10::optional<at::Tensor> val = ::c10::nullopt;
             if (vals_to_params_map.find(v) != vals_to_params_map.end()) {
-              // If the input is a parameter, insert a constant of its value as
-              // input.
-              auto val = vals_to_params_map.find(v)->second.second.toTensor();
+              val = vals_to_params_map.find(v)->second.second.toTensor();
+            } else if (ConstantValueMap::HasValue(v->debugName())) {
+              val = ConstantValueMap::GetValue(v->debugName());
+            }
+
+            if (val.has_value()) {
               return n_graph
                   ->insertNode(n_graph->create(::c10::onnx::Constant)
-                                   ->t_(attr::value, val))
+                                   ->t_(attr::value, val.value()))
```

Comment:
Just a Q: why do we call .value() here and not before when both come from `vals_to_params_map.find(v)->second.second.toTensor();`? Is val casted to ::c10::optional?

Reply:
yes, `val` explicitly declared as optional at L305.